### DISCUSSION TITLE
Sharing client between thread or one client per thread ?

### DISCUSSION THREAD

**Discussion by jjauzion** [pymodbus-dev/pymodbus#1461] [Help] [answered]
Hello,

I'm just starting using this very nice package for Modbus and I'm wondering what is the recommended design for my use case:

I have several threads that read sensors value with Modbus. 
I'm wondering if it is better to create a new client in each thread:
```
class sensor():
    def __init__(self):
        MODBUS_TCP_TIMEOUT = 2
        DEFAULT_TCP_PORT_FOR_MODBUS_GATEWAY = 512

        pymodbus.constants.Defaults.Timeout = MODBUS_TCP_TIMEOUT
        client = ModbusTcpClient(self._topology.MODBUS_CONFIG.MODBUS_GATEWAY_ADDRESSES, port=DEFAULT_TCP_PORT_FOR_MODBUS_GATEWAY)
        connection = client.connect()
        request = client.read_holding_registers(...)
```
Or if it is better to share the client:
```
class sensor():
    def __init__(self, modbus_client):
        request = modbus_client.read_holding_registers(...)
```

What do you advise ? Also would you recommend using a Async client in this use case ?
Sorry if my question is trivial, I just read the README and a couple of example so far and I will be very grateful if you point me to more documentions if there are

Thanks for your help

**Comment by janiversen** [comment, marked as answer]
First of all use async !
Secondly, your device probably do not allow many connections, but if it does then you can have 1 client pr thread.

ModbusClient is NOT threadsafe so please do not use the same client from multiple threads....instead use asyncio and get a lot better performance.

**Reply by janiversen** [reply]
If you have read the README, then you have all the pointers you need, we provide a very extensive documentation, including all legal API calls.

**Reply by jjauzion** [reply]
Many thanks for your answer, very helpful :)
